dashboard/test/lib/user_multi_auth_helper_test.rb

```diff
@@ -0,0 +1,275 @@
+      provider: nil,
+      sponsored?: false,
+      email: '',
+      hashed_email: nil,
+      username: :not_empty,
+      encrypted_password: :not_empty,
+      parent_email: :not_empty
+
+    migrate user
+
+    # A migrated parent-managed student has no authentication option rows
+    # because they sign in with username+password or word/picture, and all of
+    # these values are stored on the user row.
+    assert_user user,
+      sponsored?: false,
+      email: '',
+      hashed_email: '',
+      username: :not_empty,
+      encrypted_password: :not_empty,
+      parent_email: :not_empty,
+      primary_authentication_option: nil,
+      authentication_options: :empty
+  end
+
+  test 'convert email+password student' do
+    user = create :student
+    assert_empty user.email
+    assert_convert_email_user user
+    assert_empty user.email
+    assert_empty user.primary_authentication_option.email
+  end
+
+  test 'convert email+password teacher' do
+    user = create :teacher
+    refute_empty user.email
+    assert_convert_email_user user
+    refute_empty user.email
+    refute_empty user.primary_authentication_option.email
+  end
+
+  def assert_convert_email_user(user)
+    original_email = user.email
+    original_hashed_email = user.hashed_email
+
+    assert_user user,
+      provider: nil,
+      hashed_email: :not_empty,
+      encrypted_password: :not_empty
+
+    migrate user
+
+    assert_user user,
+      email: original_email,
+      hashed_email: original_hashed_email,
+      encrypted_password: :not_empty,
+      primary_authentication_option: {
+        credential_type: 'email',
+        authentication_id: original_hashed_email,
+        email: original_email,
+        hashed_email: original_hashed_email,
+        data: nil
+      }
+  end
+
+  #
+  # Trusted email from Oauth:
+  #
+
+  test 'convert Google OAuth teacher' do
+    user = create(:google_oauth2_teacher)
+    initial_email = user.email
+    initial_hashed_email = user.hashed_email
+    initial_authentication_id = user.uid
+    initial_oauth_token = user.oauth_token
+    initial_oauth_token_expiration = user.oauth_token_expiration
+    initial_oauth_refresh_token = user.oauth_refresh_token
+
+    assert_user user,
+      provider: 'google_oauth2',
+      email: :not_empty,
+      hashed_email: :not_empty,
+      uid: :not_nil,
+      oauth_token: :not_nil,
+      oauth_token_expiration: :not_nil,
+      oauth_refresh_token: :not_nil
+
+    migrate user
+
+    assert_user user,
+      email: initial_email,
+      hashed_email: initial_hashed_email,
+      primary_authentication_option: {
+        credential_type: 'google_oauth2',
+        authentication_id: initial_authentication_id,
+        email: initial_email,
+        hashed_email: initial_hashed_email,
+        data: {
+          oauth_token: initial_oauth_token,
+          oauth_token_expiration: initial_oauth_token_expiration,
+          oauth_refresh_token: initial_oauth_refresh_token
+        }
```

Comment:
Does this need a `.to_json`?

Reply:
Nevermind, just saw that you have special handling below. :+1:

Reply:
:+1: I tried to optimize for readability in the actual test cases, so details like that are hidden in the test helpers.